pvlib/soiling.py

```diff
@@ -76,8 +76,17 @@ def hsu(rainfall, cleaning_threshold, tilt, pm2_5, pm10,
     # cleaning is True for intervals with rainfall greater than threshold
     cleaning_times = accum_rain.index[accum_rain >= cleaning_threshold]
 
-    horiz_mass_rate = pm2_5 * depo_veloc['2_5']\
-        + np.maximum(pm10 - pm2_5, 0.) * depo_veloc['10'] * 3600
+    # determine the time intervals in seconds (dt_sec)
+    dt = rainfall.index.values       # datetimes in nanoseconds
+    dt2 = np.roll(dt.copy(), 1)       # shift array values by one place
+    # subtract shifted values from original and convert to seconds
+    dt3 = np.delete((dt-dt2)/1e9, 0)
+    # append last value to end so same number of elements in the array
+    dt_sec = np.append(dt3, dt3[-1]).astype('float64')
```

Comment:
I think this puts the time difference one time step out of sync with the rainfall.  `rainfall` is the accumulation during each time interval, so implicitly, `rainfall[0]` is the amount accumulated _before_ the first timestamp, and we don't know that interval length. Here using t to represent rainfail.index, repeating the last time difference value leads to `horiz_mass_rate[i]` being equal to (some constant) * (t_i+1 - t_i) rather than equal to constant * (t_i - t_i-1).

What about
```
dt = np.array(rainfall.index.values / 1e-9)  # datetimes in seconds
dt_sec = dt.diff()  # will have np.nan as the first value
dt_sec[0] = dt_sec[1]  # reasonable assumption that the interval prior to the first value is equal in length to the first interval
```

Reply:
I see your point, but your suggested code does not seem to be valid.  How about this solution:
`dt = np.array([(rainfall.index - rainfall.index.shift(-1)).total_seconds()])`

Reply:
I didn't test that code, obviously :) I'm guessing that using `numpy.diff` will have some performance advantage over the explicit subtracting a shift. But it probably doesn't make a whole lot of difference. @kanderso-nrel ?

Reply:
I agree with your reasoning, however your suggested code does not work in practice.  Here is what I suggest after testing several options:
 dt = rainfall.index
 dt2 = np.roll(dt.copy(),1) 
 dt3 = np.delete((dt-dt2),0).total_seconds()
 dt_sec = np.append(dt3[0],dt3)`

Reply:
Some quick timing comparisons:

```python
import numpy as np
import pandas as pd

N = 8760
data = np.arange(1, N)
idx = pd.date_range('2019-01-01', freq='h', periods=len(data))
rainfall = pd.Series(data, idx)

def f1():
    return (rainfall.index - rainfall.index.shift(-1)).total_seconds()

def f2():
    prior_ts = (rainfall.index[0] - (rainfall.index[1] - rainfall.index[0]))
    # prepend added in np 1.16.0
    dt_sec = np.diff(rainfall.index.astype(int), prepend=prior_ts.timestamp()*1e9) * 1e-9
    return dt_sec

def f3():
    first_diff = (rainfall.index[1] - rainfall.index[0]).total_seconds()
    dt_sec = np.ediff1d(rainfall.index.astype(int) * 1e-9, to_begin=first_diff)
    return dt_sec

def f4():
    first_diff = (rainfall.index[1] - rainfall.index[0]).total_seconds()*1e9
    dt_sec = np.ediff1d(rainfall.index.astype(int), to_begin=first_diff)*1e-9
    return dt_sec
```

```
%timeit f1()
403 µs ± 3.65 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

%timeit f2()
161 µs ± 6.25 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

%timeit f3()
401 µs ± 1.11 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

%timeit f4()
139 µs ± 1.34 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

`np.diff` is faster, at least on my machine, but the problem with is that it drops the unknown value instead of keeping it as `nan` so you have to kludge in a temporary value with the `prepend` argument.  However, `prepend` was only added in numpy 1.16.0, and our lowest supported version is currently 1.12.  There is an alternative method `np.ediff1d` that also shows the speed increase.  But anyway they're all well under a millisecond for an 8760, so IMO the speed isn't really relevant and code clarity is more important. 

Interestingly the timing changes dramatically whether you rescale by 1e9 before or after the diff (f3 vs f4)... strange.  I don't think it's the difference between diffing arrays with dtype=int or dtype=float.  Maybe it's somehow the difference between fitting in CPU cache or not.